checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -223,6 +253,10 @@ protected void refineGT(
             knownToBeLessThanLengthOf(array, right, store, in);
         }
 
+        if (right instanceof NumericalAdditionNode || right instanceof NumericalSubtractionNode) {
+            propagateToSubterms((BinaryOperationNode) right, refinedRight, store);
```

Comment:
I'm not sure that's the right thing to do. While the methods are similar, `knownToBeLessThanLengthOf` has some strong and different assumptions than `propagateToSubterms`. Maybe there is some way to abstract out some of the internal logic; I'll look into that.